### ISSUE TITLE
Client side identity manager

### ISSUE THREAD

**Issue by weitzhandler** [microsoft/ProjectReunion#31]
# Proposal: Client side authentication manager

## Summary
ASP.NET Identity is a magnificent tool to manage identity and authentication in the server side, including integration with EF Core, and is mature enough.

It would be complete, if there would be a client side framework, that knows how to interact with an Identity-enabled server that offers features like:

- User authentication management, via JWT tokens, claims etc.- Integration with `Thread.CurrentPrincipal`.
- Integration with `IHttpClientFactory`, to allow generating authentication-clients.
- Module authorization (e.g. add `[Authorize(Roles = RoleNames.Admin)]` attribute on a ViewModel to only enable it for admins, etc.
- Have some way to get notified when a user has been logged in / out.
- User management API that can easily interact with an ASP.NET Identity server out the box.

## Rationale
Based on my understanding, Reunion is the largest .NET client-side project, and as such, although not directly a WinUI feature request, is a client-specific feature request in the .NET.

Cross posted from [here](https://github.com/dotnet/aspnetcore/issues/21967).

**Comment by jonwis** [comment]
Have you seen https://github.com/AzureAD/microsoft-authentication-library-for-dotnet ?  What additional layers would you want on the client side?

**Comment by hansmbakker** [comment]
For what kind of .Net client apps are you looking for a solution? Blazor? Mobile? Desktop/UWP?

For UWP there is the combination of
- [`WebAccountManager`](https://docs.microsoft.com/en-us/windows/uwp/security/web-account-manager) lets you use accounts that are known to Windows 10 and lets you use authentication tokens from these accounts
- [`WebAccountProvider`](https://docs.microsoft.com/en-us/uwp/api/windows.security.credentials.webaccountprovider?view=winrt-18362) lets you implement the logic for fetching and refreshing tokens for a custom account. It also lets you set sso cookies for use in a browser or browser component.

For Microsoft (personal/work/school) accounts and for certain websites you do not need to implement a custom `WebAccountProvider`. However if your backend has its own accounts, then those will not work and you might need to write a `WebAccountProvider` yourself.

Unfortunately, there is [no good documentation](https://github.com/MicrosoftDocs/windows-uwp/issues/671) [nor is there a sample](https://github.com/microsoft/Windows-universal-samples/issues/911) how to create a custom `WebAccountProvider`, so using this technology with custom accounts is not easy.

There is only an unofficial sample, which is helpful but also has some improvement points and the sso cookies functionality seems not to work anymore.

**Comment by hansmbakker** [comment]
If your website offers OAuth, then you might be able to use https://docs.microsoft.com/en-us/windows/uwp/security/web-authentication-broker as well. However I believe that one does not integrate with the accounts registered in Windows, and it won't offer a "reusable experience".

It won't set SSO cookies for you, it won't register your account with Windows, it won't give other apps the possibility of using the same token logic.

**Comment by weitzhandler** [comment]
WOW.
Yes I meant mainly UWP. I wish this functionality has been abstracted out of UWP and became a standalone NuGet package, but this is exactly what I was looking for.
Gonna close this issue now.

**Comment by hansmbakker** [comment]
Glad that it gave you a pointer!

You can leave the issue open though, as the documentation needs to be improved for it to be more useful.

**Comment by weitzhandler** [comment]
@jonwis Yes checked it about a year ago, it doesn't support custom authentication, i.e. JWT tokens etc.

**Comment by jonwis** [comment]
> Not sure if this functionality is in scope?

Sure! One goal of Project Reunion is bringing some of the Windows Platform functionality initially developed for UWPs over to Win32 applications.

@weitzhandler  - So is the feature request then "make the web authentication broker available to Win32 apps" ?

**Comment by hansmbakker** [comment]
@jonwis I won't speak for the needs of @weitzhandler, but can you please also see what is possible for `WebAccountProvider`?
- I'm not sure that all functionality that was there still works (sso cookies set by WebAccountProvider app being injected in a browser)
- documentation is suboptimal

**Comment by weitzhandler** [comment]
Ideally there should be a .NET Standard API on NuGet that integrates with IHttpClientFactory and IPrincipal/IIdentity, that works in every type of .NET app.
Then taking it a step further and adding WinUI controls on top of it. It would be way easier to abstract it away and keep logic and UI separated.

**Comment by michael-hawker** [comment]
🦙 Cross-linking to the WebAuthenticationBroker question in the WebView2 repo [here](https://github.com/MicrosoftEdge/WebViewFeedback/issues/171)
